### PR TITLE
Enables float16 gemm on gpuarray when the cuda version supports it

### PR DESCRIPTION
**Pull request by abergeron**
This requires https://github.com/Theano/libgpuarray/pull/77.

### CODE REVIEW

**Comment by abergeron** [comment]
This can now be merged since the dependent PR is merged.

File: theano/sandbox/gpuarray/opt.py [hunk 1]
```diff
@@ -604,6 +607,34 @@ def local_gpua_gemm(node):
...
+
+@register_opt()
+@alpha_merge(GpuGemm, alpha_in=1, beta_in=2, nd=2)
+def local_gpuagemm_alpha_merge(node, *inputs):
+    return [gpugemm_no_inplace(*inputs)]
+
+
+@register_opt()
+@output_merge(GpuGemm, alpha_in=1, beta_in=2, out_in=0, nd=2)
+def local_gpuagemm_output_merge(node, *inputs):
```
Comment (nouiz) [inline theano/sandbox/gpuarray/opt.py]: this opt and the alpha merge need tests.

File: theano/sandbox/gpuarray/opt.py [hunk 2]
```diff
@@ -604,6 +607,34 @@ def local_gpua_gemm(node):
 
 
 @register_opt('fast_compile')
+@op_lifter([tensor.basic.Dot])
+def local_gpua_hgemm(node):
+    from theano.sandbox.cuda import nvcc_compiler
+    if nvcc_compiler.nvcc_version < '7.5':
+        return
```
Comment (nouiz) [inline theano/sandbox/gpuarray/opt.py]: I would move that to just before the return and print a nice user warning that updating to cuda 7.5 would speed up his computation.

File: theano/sandbox/gpuarray/tests/test_blas.py [hunk 3]
```diff
@@ -115,3 +117,47 @@ def clone(self, op):
...
+    f = theano.function([m32, m], tensor.dot(m32, m), mode=mode_with_gpu)
+    assert len([node for node in f.maker.fgraph.apply_nodes
+                if isinstance(node.op, GpuGemm)]) == 0
+
+    f = theano.function([m, m2], tensor.dot(m, m2), mode=mode_with_gpu)
+    assert len([node for node in f.maker.fgraph.apply_nodes
+                if isinstance(node.op, GpuGemm)]) == 1
+
+
+def test_hgemm_value():
```
Comment (nouiz) [inline theano/sandbox/gpuarray/tests/test_blas.py]: merge this test in the previous one to compile less Theano funcitons and speed up tests. Mostly take the last 3 code lines and put it in the previous test.

**Comment by nouiz** [comment]
Quand je compile j'ai des erreurs:

mod.cu(147): error: identifier "ga_size" is undefined
mod.cu(149): error: identifier "ga_size" is undefined
mod.cu(150): error: identifier "ga_ssize" is undefined
mod.cu(152): error: identifier "ga_size" is undefined
mod.cu(153): error: identifier "ga_ssize" is undefined

File: theano/sandbox/gpuarray/opt.py (hunk 1 above)
Comment (abergeron) [inline theano/sandbox/gpuarray/opt.py]: They are tested, but not specifically for this op.  Do you want extra tests?

**Comment by abergeron** [comment]
A rebase fixes the compilation problems.

File: theano/sandbox/gpuarray/opt.py (hunk 1 above)
Comment (nouiz) [inline theano/sandbox/gpuarray/opt.py]: We need at least test to make sure we don't comment the code.

**Comment by abergeron** [comment]
Now with more tests!

**Comment by nouiz** [comment]
there is a flake8 error:

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/sandbox/gpuarray/opt.py:614:9: F821 undefined name 'log'

**Comment by abergeron** [comment]
Should be fixed now.

**Comment by nouiz** [comment]
Seem good, we can merge when travis pass.

**Pull request merged**